sse/event_source.go

```diff
@@ -38,23 +38,47 @@ type EventSource struct {
 
 	currentReadCloser *ReadCloser
 	lastEventID       string
-	retryInterval     time.Duration
 	lock              sync.Mutex
 
-	closeOnce *sync.Once
-	closed    chan struct{}
+	closeOnce   *sync.Once
+	closed      chan struct{}
+	retryParams RetryParams
+}
+
+type RetryParams struct {
+	RetryInterval time.Duration
+	MaxRetries    uint16
 }
 
 func NewEventSource(client *http.Client, defaultRetryInterval time.Duration, requestCreator func() *http.Request) *EventSource {
+	retryParams := RetryParams{
+		RetryInterval: defaultRetryInterval,
+	}
+	return createEventSource(client, retryParams, requestCreator)
+}
+
+func createEventSource(client *http.Client, retryParams RetryParams, requestCreator func() *http.Request) *EventSource {
 	return &EventSource{
 		client:        client,
 		createRequest: requestCreator,
 
-		retryInterval: defaultRetryInterval,
+		closeOnce:   new(sync.Once),
+		closed:      make(chan struct{}),
+		retryParams: retryParams,
+	}
+}
+
+func ConnectWithParams(client *http.Client, retryParams RetryParams, requestCreator func() *http.Request) (*EventSource, error) {
```

Comment:
I was thinking of something more like how Go exposes a `*net.Dialer` that exposes fields for injecting configuration, which lets you e.g. pass `nil` for things like `RequestCreator` so it can have its own sane defaults.

https://golang.org/pkg/net/#Dialer